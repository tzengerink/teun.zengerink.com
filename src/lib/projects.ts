import fs from 'fs'
import path from 'path'
import slugify from 'slugify'
import yaml from 'js-yaml'

export enum Width {
  Mobile = 768,
  Desktop = 1280,
}

interface ConfigCaption {
  key: string
  caption: string
}

interface ConfigItem {
  title: string
  statement?: string
  isArchived?: boolean
  captions?: ConfigCaption[]
}

export interface Export {
  width: Width
  url: string
}

export interface Photo {
  key: string
  caption?: string
  exports: Export[]
}
export interface Project {
  title: string
  slug: string
  photos: Photo[]
  statement?: string
  isArchived?: boolean
}

export const MAX_WIDTHS = [Width.Mobile, Width.Desktop]

const generatePhotos = (slug: string, captions: ConfigCaption[]): Photo[] => {
  const photosDirectory = path.join(process.cwd(), 'public/photos', slug, `${Width.Mobile}w`)
  const fileNames = fs.readdirSync(photosDirectory)

  return fileNames.map((filename) => {
    const key = path.basename(filename, '.jpg')
    const caption = captions?.find((c) => c.key == key)

    return {
      key,
      caption: caption?.caption ?? '',
      exports: MAX_WIDTHS.map((width) => ({ width, url: `/photos/${slug}/${width}w/${filename}` })),
    }
  })
}

const generateProjects = (item: ConfigItem[]): Project[] => {
  return item.map((project) => {
    const slug = slugify(project.title, { lower: true })

    return {
      slug,
      title: project.title,
      statement: project.statement ?? '',
      isArchived: project.isArchived ?? false,
      photos: generatePhotos(slug, project.captions),
    }
  })
}

export const getProjects = async (): Promise<Project[]> => {
  let projects: Project[] = []

  try {
    const file = path.join(process.cwd(), 'src', 'config.yml')
    const config = yaml.load(fs.readFileSync(file, 'utf-8'))
    projects = await generateProjects(config)
  } catch (e) {
    console.error(e)
  }

  return projects
}
