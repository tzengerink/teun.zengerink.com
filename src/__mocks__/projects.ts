import { Project } from '../lib/projects'

const projects: Project[] = [
  {
    title: 'My First Project',
    slug: 'my-first-project',
    statement: 'The *statement* for the _project_',
    photos: [
      {
        key: '01',
        exports: [
          { width: 768, url: `/photos/my-first-project/768w/01.jpg` },
          { width: 1280, url: `/photos/my-first-project/1280w/01.jpg` },
        ],
      },
      {
        key: '02',
        exports: [
          { width: 768, url: `/photos/my-first-project/768w/02.jpg` },
          { width: 1280, url: `/photos/my-first-project/1280w/02.jpg` },
        ],
      },
      {
        key: '03',
        exports: [
          { width: 768, url: `/photos/my-first-project/768w/03.jpg` },
          { width: 1280, url: `/photos/my-first-project/1280w/03.jpg` },
        ],
      },
    ],
  },
  {
    title: 'Project without a statement',
    slug: 'no-statement',
    photos: [
      {
        key: '01',
        caption: 'This is _image 1_',
        exports: [
          { width: 768, url: `/photos/no-statement/768w/01.jpg` },
          { width: 1280, url: `/photos/no-statement/1280w/01.jpg` },
        ],
      },
      {
        key: '02',
        caption: 'This is _image 2_',
        exports: [
          { width: 768, url: `/photos/no-statement/768w/02.jpg` },
          { width: 1280, url: `/photos/no-statement/1280w/02.jpg` },
        ],
      },
      {
        key: '03',
        caption: 'This is _image 3_',
        exports: [
          { width: 768, url: `/photos/no-statement/768w/03.jpg` },
          { width: 1280, url: `/photos/no-statement/1280w/03.jpg` },
        ],
      },
      {
        key: '04',
        caption: 'This is _image 4_',
        exports: [
          { width: 768, url: `/photos/no-statement/768w/04.jpg` },
          { width: 1280, url: `/photos/no-statement/1280w/04.jpg` },
        ],
      },
    ],
  },
  {
    title: 'Archived project',
    slug: 'archived-project',
    isArchived: true,
    photos: [
      {
        key: '01',
        exports: [
          { width: 768, url: `/photos/archived-project/768w/01.jpg` },
          { width: 1280, url: `/photos/archived-project/1280w/01.jpg` },
        ],
      },
      {
        key: '02',
        exports: [
          { width: 768, url: `/photos/archived-project/768w/02.jpg` },
          { width: 1280, url: `/photos/archived-project/1280w/02.jpg` },
        ],
      },
    ],
  },
]

export default projects
