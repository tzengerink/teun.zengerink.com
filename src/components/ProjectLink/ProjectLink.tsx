import classNames from 'classnames'
import Link from 'next/link'
import React from 'react'
import { Project } from '../../lib/types'
import Photo from '../Photo/Photo'

interface ProjectLink {
  project: Project
}

const ProjectLink: React.FC<ProjectLink> = ({ project }) => {
  return (
    <div className={classNames('h-[calc(100vh-150px)]', 'md:h-[100vh]')}>
      <Link href={`/work/${project?.slug}/${project?.photos[1].key}`}>
        <div
          className={classNames(
            'group',
            'relative',
            'inline-block',
            'leading-[0px]',
            'cursor-pointer',
            'top-1/2',
            'translate-y-[-50%]',
            'md:left-1/2',
            'md:translate-x-[-50%]',
          )}
        >
          <div
            className={classNames(
              'absolute',
              'inset-0',
              'top-1/2',
              'opacity-0',
              'group-hover:opacity-100',
              'text-center',
              'text-white',
              'z-20',
              'after:content-[""]',
              'after:absolute',
              'after:top-[-100%]',
              'after:bottom-0',
              'after:left-0',
              'after:right-0',
              'after:bg-black/40',
              'after:z-10',
            )}
          >
            {project?.title}
          </div>
          <Photo photo={project?.photos[0]} alt={project?.title} />
        </div>
      </Link>
    </div>
  )
}

export default ProjectLink
