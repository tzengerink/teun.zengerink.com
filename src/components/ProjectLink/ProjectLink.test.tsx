import { render } from '@testing-library/react'
import React from 'react'
import mockLink from '../../__mocks__/nextLink'
import projects from '../../__mocks__/projects'
import ProjectLink from './ProjectLink'

jest.mock('next/link', () => mockLink)

describe('SinglePhoto', () => {
  it('renders correctly for projects with a statement', () => {
    const { container } = render(<ProjectLink project={projects[0]} />)
    expect(container).toMatchSnapshot()
  })

  it('renders correctly for projects without a statement', () => {
    const { container } = render(<ProjectLink project={projects[1]} />)
    expect(container).toMatchSnapshot()
  })
})
