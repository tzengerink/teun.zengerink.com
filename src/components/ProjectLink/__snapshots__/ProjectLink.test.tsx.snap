// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SinglePhoto renders correctly for projects with a statement 1`] = `
<div>
  <a
    href="/work/my-first-project/02"
  >
    <div
      class="group relative inline-block leading-[0px] cursor-pointer md:top-1/2 md:left-1/2 md:translate-x-[-50%] md:translate-y-[-50%]"
    >
      <div
        class="absolute inset-0 top-1/2 opacity-0 group-hover:opacity-100 text-center text-white z-20 after:content-[\\"\\"] after:absolute after:top-[-100%] after:bottom-0 after:left-0 after:right-0 after:bg-black/40 after:z-10"
      >
        My First Project
      </div>
      <img
        alt="My First Project"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh]"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/my-first-project/1280w/01.jpg"
        srcset="/photos/my-first-project/768w/01.jpg 768w, /photos/my-first-project/1280w/01.jpg 1280w"
      />
    </div>
  </a>
</div>
`;

exports[`SinglePhoto renders correctly for projects without a statement 1`] = `
<div>
  <a
    href="/work/no-statement/02"
  >
    <div
      class="group relative inline-block leading-[0px] cursor-pointer md:top-1/2 md:left-1/2 md:translate-x-[-50%] md:translate-y-[-50%]"
    >
      <div
        class="absolute inset-0 top-1/2 opacity-0 group-hover:opacity-100 text-center text-white z-20 after:content-[\\"\\"] after:absolute after:top-[-100%] after:bottom-0 after:left-0 after:right-0 after:bg-black/40 after:z-10"
      >
        Project without a statement
      </div>
      <img
        alt="Project without a statement"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh]"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/no-statement/1280w/01.jpg"
        srcset="/photos/no-statement/768w/01.jpg 768w, /photos/no-statement/1280w/01.jpg 1280w"
      />
    </div>
  </a>
</div>
`;
